core/src/main/java/org/apache/cloudstack/direct/download/HttpsDirectTemplateDownloader.java

```diff
@@ -0,0 +1,252 @@
+                .setConnectTimeout(connectTimeout == null ? 5000 : connectTimeout)
+                .setConnectionRequestTimeout(connectionRequestTimeout == null ? 5000 : connectionRequestTimeout)
+                .setSocketTimeout(soTimeout == null ? 5000 : soTimeout).build();
+        httpsClient = HttpClients.custom().setSSLSocketFactory(factory).setDefaultRequestConfig(config).build();
+        createUriRequest(url, headers);
+        String downloadDir = getDirectDownloadTempPath(templateId);
+        File tempFile = createTemporaryDirectoryAndFile(downloadDir);
+        setDownloadedFilePath(tempFile.getAbsolutePath());
+    }
+
+    protected void createUriRequest(String downloadUrl, Map<String, String> headers) {
+        req = new HttpGet(downloadUrl);
+        if (MapUtils.isNotEmpty(headers)) {
+            for (String headerKey: headers.keySet()) {
+                req.setHeader(headerKey, headers.get(headerKey));
+            }
+        }
+    }
+
+    private SSLContext getSSLContext() {
+        try {
+            KeyStore customKeystore  = KeyStore.getInstance("jks");
+            try (FileInputStream instream = new FileInputStream(new File("/etc/cloudstack/agent/cloud.jks"))) {
+                String privatePasswordFormat = "sed -n '/keystore.passphrase/p' '%s' 2>/dev/null  | sed 's/keystore.passphrase=//g' 2>/dev/null";
+                String privatePasswordCmd = String.format(privatePasswordFormat, "/etc/cloudstack/agent/agent.properties");
+                String privatePassword = Script.runSimpleBashScript(privatePasswordCmd);
+                customKeystore.load(instream, privatePassword.toCharArray());
+            }
+            KeyStore defaultKeystore = KeyStore.getInstance(KeyStore.getDefaultType());
+            String relativeCacertsPath = "/lib/security/cacerts".replace("/", File.separator);
+            String filename = System.getProperty("java.home") + relativeCacertsPath;
+            try (FileInputStream is = new FileInputStream(filename)) {
+                String password = "changeit";
+                defaultKeystore.load(is, password.toCharArray());
+            }
+            TrustManager[] tm = HttpsMultiTrustManager.getTrustManagersFromKeyStores(customKeystore, defaultKeystore);
+            SSLContext sslContext = SSLUtils.getSSLContext();
+            sslContext.init(null, tm, null);
+            return sslContext;
+        } catch (KeyStoreException | NoSuchAlgorithmException | CertificateException | IOException | KeyManagementException e) {
+            s_logger.error(String.format("Failure getting SSL context for HTTPS downloader, using default SSL context: %s", e.getMessage()), e);
+            try {
+                return SSLContext.getDefault();
+            } catch (NoSuchAlgorithmException ex) {
+                throw new CloudRuntimeException(String.format("Cannot return the default SSL context due to: %s", ex.getMessage()), e);
+            }
+        }
```

Comment:
as far as I understand, these are the main changes, right ? @nvazquez

Reply:
Yes @weizhouapache - then have also refactored the downloaders

Reply:
cool. thanks @nvazquez 
code lgtm